zkevm-circuits/src/util/int_decomposition.rs

```diff
@@ -0,0 +1,100 @@
+//! Define IntDecomposition to decompose int into byte limbs
+use eth_types::{Field, ToLittleEndian, H160, U256};
+use gadgets::util::{sum, Expr};
+use halo2_proofs::{
+    circuit::{AssignedCell, Value},
+    plonk::{Error, Expression},
+};
+use itertools::Itertools;
+
+use crate::evm_circuit::{
+    param::N_BYTES_HALF_WORD,
+    util::{rlc, CachedRegion, Cell},
+};
+
+use super::word::{Word, WordExpr};
+
+#[derive(Clone, Debug)]
+/// IntDecomposition decompose integer into byte limbs
+pub struct IntDecomposition<F, const N_LIMBS: usize> {
+    /// inner cells in little-endian for synthesis
+    pub limbs: [Cell<F>; N_LIMBS],
+}
+
+impl<F: Field, const N_LIMBS: usize> IntDecomposition<F, N_LIMBS> {
+    /// new by cell limbs
+    pub fn new(limbs: [Cell<F>; N_LIMBS]) -> Self {
+        Self { limbs }
+    }
+
+    /// assign bytes to cells
+    pub fn assign<const N_LIMBS_ASSIGN: usize>(
+        &self,
+        region: &mut CachedRegion<'_, '_, F>,
+        offset: usize,
+        bytes: Option<[u8; N_LIMBS_ASSIGN]>,
+    ) -> Result<Vec<AssignedCell<F, F>>, Error> {
+        assert!(N_LIMBS_ASSIGN >= N_LIMBS);
+        bytes.map_or(Err(Error::Synthesis), |bytes| {
+            self.limbs
+                .iter()
+                .zip(bytes.iter())
+                .map(|(cell, byte)| {
+                    cell.assign(region, offset, Value::known(F::from(*byte as u64)))
+                })
+                .collect()
+        })
+    }
+
+    /// assign h160 to cells
+    pub fn assign_h160(
+        &self,
+        region: &mut CachedRegion<'_, '_, F>,
+        offset: usize,
+        h160: H160,
+    ) -> Result<Vec<AssignedCell<F, F>>, Error> {
+        let mut bytes = *h160.as_fixed_bytes();
+        bytes.reverse();
+        self.assign(region, offset, Some(bytes))
+    }
+
+    /// assign u256 to cells
+    pub fn assign_u256(
+        &self,
+        region: &mut CachedRegion<'_, '_, F>,
+        offset: usize,
+        u256: U256,
+    ) -> Result<Vec<AssignedCell<F, F>>, Error> {
+        self.assign(region, offset, Some(u256.to_le_bytes()))
+    }
+
+    /// assign all limbs into one expression
+    pub fn sum_expr(&self) -> Expression<F> {
+        sum::expr(self.limbs.clone())
+    }
+}
+
+impl<F: Field, const N_LIMBS: usize> Expr<F> for IntDecomposition<F, N_LIMBS> {
+    fn expr(&self) -> Expression<F> {
+        rlc::expr(&self.limbs.clone().map(|limb| limb.expr()), 256.expr())
```

Comment:
I would suggest doing `assert!(N_LIMBS <= 31)` as a safety check here.  This is because for N_LIMBS > 31 we overflow the last byte in the field.

Reply:
Added in this commit https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1411/commits